resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/v2/KubernetesSparkDependencyServiceImpl.scala

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.deploy.rest.kubernetes.v2
+
+import java.io.{File, FileOutputStream, InputStream, OutputStream}
+import java.security.SecureRandom
+import java.util.UUID
+import javax.ws.rs.core.StreamingOutput
+
+import com.google.common.io.{BaseEncoding, ByteStreams, Files}
+import scala.collection.mutable
+
+import org.apache.spark.SparkException
+import org.apache.spark.deploy.rest.KubernetesCredentials
+import org.apache.spark.util.Utils
+
+private[spark] class KubernetesSparkDependencyServiceImpl(dependenciesRootDir: File)
+    extends KubernetesSparkDependencyService {
+
+  private val DIRECTORIES_LOCK = new Object
+  private val REGISTERED_DRIVERS_LOCK = new Object
+  private val SPARK_APPLICATION_DEPENDENCIES_LOCK = new Object
+  private val SECURE_RANDOM = new SecureRandom()
+  // TODO clean up these resources based on the driver's lifecycle
+  private val registeredDrivers = mutable.Set.empty[PodNameAndNamespace]
```

Comment:
for now, using an LRU cache with maximum size and expireAfterWrite of about 7 days and expireAfterRead of about 24 hours at least prevents OOMs in this service

We'll need to rethink this for streaming and other long-running applications though